### DISCUSSION TITLE
Time issues

### DISCUSSION THREAD

**Discussion by OOHehir** [LIFsCode/ELOC-3.0#92] [General]
@EDsteve @LIFsCode There's quite a few time issues going on & I'm afraid I don't have a handle where one begins & another ends, what the interaction is & where I can help.

From what I can tell the current time issues are:

1. Issue #86, this seems to be originating on the Android side, is there any fix required on the ELOC firmware?
2. Issue #84, incorrect timestamps -> Do we believe this overlaps #77? I.e. if the ESP32 system time is off I believe its also likely the timestamps will also be incorrect as they use system time (via ESP32Time). Are the sound files of expected length & free of sound distortion?  Or is there an issue with the sound recordings?
3. Issue #77, the 'uptime' in the status report uses the espidf function esp_timer_get_time() so do we believe the only fix for this is an external crystal for improved time stability or something else? As above, if the uptime is not correct this points to the ELOC system time being incorrect & so it's likely any file timestamps will be incorrect.

Is there also SD card issues associated with the above?

**Comment by OOHehir** [comment]
@EDsteve @LIFsCode Following on from above when obtaining time the ELOC has the following sequence:
BT connection -> cmd_SetTime() ->  this calls ESP32Time::SetTime() -> this calls settimeofday();

I see issues reported with [settimeofday](https://github.com/espressif/arduino-esp32/issues/2526) & also [here](https://stackoverflow.com/questions/74698406/esp32-s2-reading-wrong-time-after-manual-set-up). I'm not certain how this might fit into the all the issues but could be one part of the problem.

**Comment by LIFsCode** [comment]
@OOHehir @EDsteve yes the situation is really confusing at the moment @OOHehir you already made a pretty good summary:

1. Yes you're right #86 is firstly an App issue. It's possible that there is a firmware issue as well but as long as the app is sending too many commands and thus causing potential data loss in the command string this hides everything else. So there is nothing to do for us right now.

2. I personally would guess #84 is more overlapping with #86. Notwithstanding the above this is definitely a secondary issue so this is probably easier to track once the other issues are solved/understood.
- @OOHehir great find with the [settimeofday](https://stackoverflow.com/questions/74698406/esp32-s2-reading-wrong-time-after-manual-set-up) that could match with #84 or #86. 


3. #77 is definitely a firmware/hardware issue. Possible root causes are
- accuracy of internal 8.5MHz. I couldn't find anything about it. @OOHehir do you have any hints on its accuracy or at least how it's generated
- any potential issues within esp libraries which causes a corruption of timestamp timer/counter @OOHehir you already found some bug references so maybe that's a good way for you to go
- I myself check on the HW settings and try to figure out a way to identify if that's the problem

**Comment by EDsteve** [comment]
@OOHehir 
> Are the sound files of expected length & free of sound distortion? Or is there an issue with the sound recordings?

Let's say i have recorded single a 12h file.
Time-length showing in Audacity. It shows the expected length of 12h.
If you are asking about the time difference between file created and file modified. Then it does not have 12h. It's off between a few seconds to a minute and can be shorter or longer than the actual 12h. @LIFsCode explaned it [here](https://github.com/LIFsCode/ELOC-3.0/issues/77#issuecomment-1984108799)
There is no issues with distortion. I only see one single wav buffer overrun in the log between each files. But that is not a big issue.

Is there also SD card issues associated with the above?
I had a suspicion. But it doesn't seem like it.

**Comment by OOHehir** [comment]
@LIFsCode @EDsteve Thanks, I'll try working on #77 first. I'll make any comments there to keep it organized.

**Comment by EDsteve** [comment]
@LIFsCode @OOHehir 
Was badly ill after my trip. So there was not much testing going on lately on my side. But out of bed and already back to 70% today. Hope that didn't happen to you as well @OOHehir :)

I will dare to update the time-problem status, So we can hopefully continue with more fun stuff :)

1. https://github.com/LIFsCode/ELOC-3.0/issues/86 Should be sovled from the App side. Haven't seen that issue (or too many requests-issue) since Joseph send me his app-update. Already on Google Play.
2. I also had the feeling it is connected to https://github.com/LIFsCode/ELOC-3.0/issues/86 So we just see and hope it doesn't happen again. For now :)
[settimeofday](https://stackoverflow.com/questions/74698406/esp32-s2-reading-wrong-time-after-manual-set-up) - Should we implement this fix or wait and see what happens?
3. https://github.com/LIFsCode/ELOC-3.0/issues/77 should work well enough. As long the times are not off a few days as they were before. We should be OK.

I will start a few long term tests with the latest commit and report back as usual?

**Reply by OOHehir** [reply]
@EDsteve good to hear you are back on you're feet! I've been away myself but will hopefully close off #77 in the next day or two..

**Comment by OOHehir** [comment]
@EDsteve @LIFsCode Can we close #77 & #84?

**Reply by LIFsCode** [reply]
@EDsteve @OOHehir  I guess so it seems understood and solved

**Comment by OOHehir** [comment]
@EDsteve @LIFsCode It would be nice to be able to understand the reason that esp_timer_get_time() is not accurate but I close this for now.
